src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -449,147 +609,129 @@ void TaskSketcherElements::onSelectionChanged(const Gui::SelectionChanges& msg)
-                previouslySelectedItemIndex>=0){
-              // select the element of the Item
-                      switch(element){
-                  case 0:
-                      ite->isLineSelected=true;
-                      break;
-                  case 1:
-                      ite->isStartingPointSelected=true;
-                      break;
-                  case 2:
-                      ite->isEndPointSelected=true;
-                      break;
-                  case 3:
-                      ite->isMidPointSelected=true;
-                      break;
-                }
+        if (item == itf) {
+
+            if (itemData->GeometryType == Part::GeomPoint::getClassTypeId()) {
+                itemData->isStartingPointSelected = !itemData->isStartingPointSelected;
+            }
+            else if (itemData->clickedOn == ElementData::ClickedOn::mid
+                && (itemData->GeometryType == Part::GeomArcOfCircle::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfEllipse::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfHyperbola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfParabola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomCircle::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomEllipse::getClassTypeId())) {
+                itemData->isMidPointSelected = !itemData->isMidPointSelected;
+            }
+            else if (itemData->clickedOn == ElementData::ClickedOn::start &&
+                (itemData->GeometryType == Part::GeomArcOfCircle::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfEllipse::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfHyperbola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfParabola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomLineSegment::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomBSplineCurve::getClassTypeId())) {
+                itemData->isStartingPointSelected = !itemData->isStartingPointSelected;
+            }
+            else if (itemData->clickedOn == ElementData::ClickedOn::end &&
+                (itemData->GeometryType == Part::GeomArcOfCircle::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfEllipse::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfHyperbola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomArcOfParabola::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomLineSegment::getClassTypeId()
+                    || itemData->GeometryType == Part::GeomBSplineCurve::getClassTypeId())) {
+                itemData->isEndPointSelected = !itemData->isEndPointSelected;

```

Comment:
I don't see why all this isn't managed at item level.

Reply:
Why should it be? Besides where would you put it exactly?
A function in ElementData?

Reply:
Because it seems really related to item state, and not to list. ;)